src/indexable.cr

```diff
@@ -170,6 +170,311 @@ module Indexable(T)
+  #   puts "#{n}. #{name}"
+  # end
+  # ```
+  #
+  # Prints
+  #
+  # ```text
+  # 1. Alice
+  # 2. Alice
+  # 3. Alice
+  # 1. Bob
+  # 2. Bob
+  # 3. Bob
+  # ```
+  def each_product(*others : Indexable, &block)
+    Indexable.each_product_impl(self, *others) { |v| yield v }
+  end
+
+  protected def self.each_product_impl(*indexables : *U, &block) forall U
+    lens = indexables.map &.size
+    return if lens.any? &.zero?
+
+    n = indexables.size
+    indices = Array.new(n, 0)
+    indices[-1] -= 1
+
+    while true
+      i = n - 1
+      indices[i] += 1
+
+      while indices[i] >= lens[i]
+        indices[i] = 0
+        i -= 1
+        return if i < 0
+        indices[i] += 1
+      end
+
+      {% begin %}
+        yield Tuple.new(
+          {% for i in 0...U.size %}
+            indexables[{{ i }}].unsafe_fetch(indices[{{ i }}]),
+          {% end %}
+        )
+      {% end %}
+    end
+  end
+
+  # Yields each ordered combination of the elements taken from each of the
+  # *indexables* as `Array`s.
+  # Traversal of elements starts from the last `Indexable`.
+  #
+  # `#each_product` is preferred over this class method when the quantity of
+  # *indexables* is known in advance.
+  #
+  # ```
+  # Indexable.each_product([%w[Alice Bob Carol], [1, 2]]) do |name, n|
+  #   puts "#{n}. #{name}"
+  # end
+  # ```
+  #
+  # Prints
+  #
+  # ```text
+  # 1. Alice
+  # 2. Alice
+  # 1. Bob
+  # 2. Bob
+  # 1. Carol
+  # 2. Carol
+  # ```
+  #
+  # By default, a new `Array` is created and yielded for each combination.
+  #
+  # * If *reuse* is an `Array`, it will be reused
+  # * If *reuse* is truthy, the method will create a new `Array` and reuse it
+  # * If *reuse* is falsey, no `Array`s will be reused.
+  #
+  # This can be used to prevent many memory allocations when each combination of
+  # interest is to be used in a read-only fashion.
+  def self.each_product(indexables : Indexable(Indexable), reuse = false, &block)
+    return if indexables.empty?
+    lens = indexables.map &.size
+    return if lens.any? &.==(0)
+
+    n = indexables.size
+    pool = Array.new(n) { |i| indexables.unsafe_fetch(i).first }
+    indices = Array.new(n, 0)
+    reuse = check_reuse(pool, n, reuse)
+
+    while true
+      yield pool_slice(pool, n, reuse)
+
+      i = n - 1
+      indices[i] += 1
+
+      while indices[i] >= lens[i]
+        indices[i] = 0
+        pool[i] = indexables[i].unsafe_fetch(0)
+        i -= 1
+        return if i < 0
+        indices[i] += 1
+      end
+
+      pool[i] = indexables[i].unsafe_fetch(indices[i])
+    end
+  end
+
+  protected def self.check_reuse(pool, size, reuse)
+    if reuse
+      unless reuse.is_a?(Array)
+        reuse = typeof(pool).new(size)
+      end
+    else
+      reuse = nil
+    end
+    reuse
+  end
+
+  protected def self.pool_slice(pool, size, reuse)
+    if reuse
+      reuse.clear
+      size.times { |i| reuse << pool[i] }
+      reuse
+    else
+      pool[0, size]
+    end
+  end
```

Comment:
These are essentially the same as the ones in `Array`, which are used by `#each_permutation`, `#each_combination`, and `#each_repeated_combination`; it follows that these methods (and their non-enumerating variants) can also be moved to `Indexable` (same goes for `#each_repeated_permutation` which uses the Cartesian product indirectly).